src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs

```diff
@@ -85,35 +85,40 @@ public virtual void Invoke(CancellationToken cancellationToken)
 
             using (new CaretPositionRestorer(this.SubjectBuffer, this.EditHandler.AssociatedViewService))
             {
-                var extensionManager = this.Workspace.Services.GetService<IExtensionManager>();
-                extensionManager.PerformAction(Provider, () =>
-                {
-                    IEnumerable<CodeActionOperation> operations = null;
+                Func<Document> getFromDocument = () => this.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                InvokeCore(getFromDocument, cancellationToken);
+            }
+        }
 
-                    // NOTE: As mentoned above, we want to avoid computing the operations on the UI thread.
-                    // However, for CodeActionWithOptions, GetOptions() might involve spinning up a dialog
-                    // to compute the options and must be done on the UI thread.
-                    var actionWithOptions = this.CodeAction as CodeActionWithOptions;
-                    if (actionWithOptions != null)
-                    {
-                        var options = actionWithOptions.GetOptions(cancellationToken);
-                        if (options != null)
-                        {
-                            operations = GetOperationsAsync(actionWithOptions, options, cancellationToken).WaitAndGetResult(cancellationToken);
-                        }
-                    }
-                    else
-                    {
-                        operations = GetOperationsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
-                    }
+        public void InvokeCore(Func<Document> getFromDocument, CancellationToken cancellationToken)
```

Comment:
well yes, sure but I didn't know if that would have any side effects, so I retained the existing codeflow.